analyzer/src/main/kotlin/managers/NPM.kt

```diff
@@ -120,17 +124,39 @@ open class NPM(analyzerConfig: AnalyzerConfiguration, repoConfig: RepositoryConf
                 NPM(analyzerConfig, repoConfig)
     }
 
-    protected open val recognizedLockFiles = NPM_LOCK_FILES
+    protected lateinit var projectRoot: File
+    protected lateinit var workspaceMatchers: List<PathMatcher>
+
+    protected fun isManagedByWorkspace(projectDir: File) = workspaceMatchers.any { it.matches(projectDir.toPath()) }
+
+    protected open fun hasLockFile(projectDir: File) = NPM_LOCK_FILES.any { projectDir.resolve(it).isFile }
+
+    protected fun hasYarnLockFile(projectDir: File, considerWorkspaces: Boolean = true) =
+            YARN_LOCK_FILES.any { lockFile ->
+                projectDir.resolve(lockFile).isFile ||
+                        (considerWorkspaces && projectRoot.resolve(lockFile).isFile && isManagedByWorkspace(projectDir))
+            }
 
     override fun command(workingDir: File?) = if (OS.isWindows) "npm.cmd" else "npm"
 
     override fun getVersionRequirement(): Requirement = Requirement.buildNPM("5.5.* - 6.4.*")
 
-    override fun mapDefinitionFiles(definitionFiles: List<File>) =
-            // Only keep those definition files that are not accompanied by a Yarn lock file.
-            definitionFiles.filterNot { definitionFile ->
-                YARN_LOCK_FILES.any { definitionFile.resolveSibling(it).isFile }
-            }
+    override fun mapDefinitionFiles(definitionFiles: List<File>): List<File> {
+        projectRoot = if (definitionFiles.count() > 1) {
+            getCommonFilePrefix(definitionFiles)
+        } else {
+            definitionFiles.first().parentFile
+        }
```

Comment:
If a repository contains two independent workspaces, or a package.json unrelated to the workspace in a separate directory, this would not work, right? What about instead checking all package.json files if they contain a "workspace" element, and building the workspace matchers based on that?

Reply:
Correct, handling of multiple workspaces is not supported (yet). However, non-workspace `package.json` NPM or Yarn projects should work: In case of an individual NPM project, `package.json` would be kept here as it's not accompanied by a Yarn lock file. And in the Yarn class, an individual Yarn project would be kept as it's accompanied by a Yarn lock file by (our) definition of a Yarn project.